File: API/Entities/Delivery.cs
using API.Entities.Enums;
using API.Entities.Interfaces;

namespace API.Entities;

public class Delivery: IEntityDate
{
    public int Id { get; set; }
    public DeliveryState State { get; set; }
    
    public int UserId { get; set; }
    public User From { get; set; }
    public Client Recipient { get; set; }
    
    public string Message { get; set; }
    public IList<SystemMessage> SystemMessages { get; set; }
    public IList<DeliveryLine> Lines { get; set; }

    public DateTime CreatedUtc { get; set; }
    public DateTime LastModifiedUtc { get; set; }
}

File: UI/Web/src/app/_models/delivery.ts
import {User} from './user';
import {Client} from './client';

export type Delivery = {
  id: number,
  state: DeliveryState,

  fromId: number,
  from?: User,

  clientId: number,
  recipient?: Client,

  message: string,
  systemMessages: SystemMessage[],
  lines: DeliveryLine[],

  createdUtc?: Date,
  lastModifiedUtc?: Date,
}

export type SystemMessage = {
  message: string,
  createdUtc: Date,
}

export type DeliveryLine = {
  productId: number,
  quantity: number,
}

export enum DeliveryState {
  InProgress = 0,
  Completed = 1,
  Handled = 2,
  Cancelled = 3,
}

export const AllDeliveryStates: DeliveryState[] = Object.values(DeliveryState).filter(
  (v): v is DeliveryState => typeof v === 'number'
);
